### PR TITLE
Add Azure managed identity support

### PR DESCRIPTION
**Pull request by metamben**
This builds on #44261 implementing milestone 1 of the epic #43348 and implements milestones 2 and 3.

Full testing requires access to an Azure environment, which @luizarakaki can provide. Without that, the UI part can be tested by adding a new Postgres database to Metabase.

Azure Managed Identity UI

image

OAuth UI

image

### CODE REVIEW

File: frontend/src/metabase/databases/components/DatabaseHostnameSectionField/DatabaseHostameSectionField.tsx [hunk 1]
```diff
@@ -22,7 +22,7 @@ const DatabaseHostnameSectionField = ({
   return (
     <FormField>
       <SectionButton type="button" onClick={handleClick}>
-        {value ? t`Use hostname` : t`Use  account name`}
+        {value ? t`Use account name` : t`Use hostname`}
```
Comment (metamben) [inline frontend/src/metabase/databases/components/DatabaseHostnameSectionField/DatabaseHostameSectionField.tsx:25]: This is an unrelated bugfix requested by @luizarakaki.

File: src/metabase/driver.clj [hunk 2]
```diff
@@ -966,6 +966,29 @@
...
+
+(defmethod incorporate-auth-provider-details :sql-jdbc
+  [_driver auth-provider auth-provider-response details]
+  (case auth-provider
+    (:oauth :azure-managed-identity)
+    (let [{:keys [access_token expires_in]} auth-provider-response]
+      (cond-> (assoc details :password access_token)
+        expires_in (assoc :password-expiry-timestamp (+ (System/currentTimeMillis)
+                                                        ;; renew the password a minute before expiry
+                                                        (* (- (parse-long expires_in) 60) 1000)))))
```
Comment (metamben) [inline src/metabase/driver.clj:988]: Ideally, this `60` should be the same as `metabase.db.data-source/azure-auth-token-renew-slack-seconds`. I'm open to suggestion where this constant should live.

File: src/metabase/models/database.clj [hunk 3]
```diff
@@ -173,14 +173,16 @@
...
+               driver
+               (m/update-existing-in db [:details :auth-provider] keyword))))]
     (cond-> database
       ;; TODO - this is only really needed for API responses. This should be a `hydrate` thing instead!
       (driver.impl/registered? driver)
       (assoc :features (driver.u/features driver (t2.realize/realize database)))
 
       (and (driver.impl/registered? driver)
-           (:details database)
+           (map? (:details database))
```
Comment (metamben) [inline src/metabase/models/database.clj:185]: All existing code seems to be able to normalize a map.

File: src/metabase/driver.clj (hunk 2 above)
Comment (lbrdnk) [reply src/metabase/driver.clj:988]: As that is something common for `driver` and `data-source` modules, why not to put it into `auth-provider`? I believe there's defined functionality common for the auth providers (`fetch-auth` defmulti), but also provider specific functionality (`fetch-auth :azure-managed-identity`).

File: src/metabase/driver/util.clj [hunk 4]
```diff
@@ -659,3 +660,19 @@
...
       (into default-sensitive-fields (map (comp keyword :name) password-fields)))
     default-sensitive-fields))
+
+(defn fetch-and-incorporate-auth-provider-details
+  "Incorporates auth-provider responses with db-details.
+
+  If you have a database you need to pass the database-id as some providers will need to save the response (e.g. refresh-tokens)."
+  ([driver db-details]
+   (fetch-and-incorporate-auth-provider-details driver nil db-details))
+  ([driver database-id {:keys [auth-provider] :as db-details}]
```
Comment (lbrdnk) [inline src/metabase/driver/util.clj:670]: The `auth-provider` is set in admin panel for user added dw. If we disable auth provider wanting to switch back to other auth method, will this value get cleaned up? If not, I _have a feeling_ it could lead to unwanted behavior.

**Review by lbrdnk — APPROVED** [review]
👍, but there are still few failures to fix. Please also see my comments prior to merging.
